core/src/main/java/org/apache/brooklyn/enricher/stock/PercentageEnricher.java

```diff
@@ -0,0 +1,127 @@
+
+    public static final ConfigKey<AttributeSensor<? extends Number>> SOURCE_CURRENT_SENSOR = ConfigKeys.newConfigKey(new TypeToken<AttributeSensor<? extends Number>>() {
+    }, "enricher.sourceCurrentSensor");
+
+    public static final ConfigKey<AttributeSensor<? extends Number>> SOURCE_TOTAL_SENSOR = ConfigKeys.newConfigKey(new TypeToken<AttributeSensor<? extends Number>>() {
+    }, "enricher.sourceTotalSensor");
+
+    public static final ConfigKey<AttributeSensor<Double>> TARGET_SENSOR = ConfigKeys.newConfigKey(new TypeToken<AttributeSensor<Double>>() {
+    }, "enricher.targetSensor");
+
+    public static final ConfigKey<Entity> PRODUCER = ConfigKeys.newConfigKey(Entity.class, "enricher.producer");
+
+    protected AttributeSensor<? extends Number> sourceCurrentSensor;
+    protected AttributeSensor<? extends Number> sourceTotalSensor;
+    protected AttributeSensor<Double> targetSensor;
+    protected Entity producer;
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+
+        this.sourceCurrentSensor = getConfig(SOURCE_CURRENT_SENSOR);
+        this.sourceTotalSensor = getConfig(SOURCE_TOTAL_SENSOR);
+        this.targetSensor = getConfig(TARGET_SENSOR);
+        this.producer = getConfig(PRODUCER) == null ? entity : getConfig(PRODUCER);
+
+        if (sourceCurrentSensor == null) {
+            throw new IllegalArgumentException("Enricher " + JavaClassNames.simpleClassName(this) + " has no " + SOURCE_CURRENT_SENSOR.getName());
+        }
+        if (sourceTotalSensor == null) {
+            throw new IllegalArgumentException("Enricher " + JavaClassNames.simpleClassName(this) + " has no " + SOURCE_TOTAL_SENSOR.getName());
+        }
+        if (targetSensor == null) {
+            throw new IllegalArgumentException("Enricher " + JavaClassNames.simpleClassName(this) + " has no " + TARGET_SENSOR.getName());
+        }
+
+        if (targetSensor.equals(sourceCurrentSensor)) {
+            throw new IllegalArgumentException("Enricher " + JavaClassNames.simpleClassName(this) + " detect cycle with " + SOURCE_CURRENT_SENSOR.getName());
+        }
+        if (targetSensor.equals(sourceTotalSensor)) {
+            throw new IllegalArgumentException("Enricher " + JavaClassNames.simpleClassName(this) + " detect cycle with " + SOURCE_TOTAL_SENSOR.getName());
+        }
+
+        subscriptions().subscribe(MutableMap.of("notifyOfInitialValue", true), producer, sourceCurrentSensor, this);
+        subscriptions().subscribe(MutableMap.of("notifyOfInitialValue", true), producer, sourceTotalSensor, this);
+
+    }
+
+    @Override
+    public void onEvent(SensorEvent<Number> event) {
+        Number current = producer.sensors().get(sourceCurrentSensor);
+        Number total = producer.sensors().get(sourceTotalSensor);
+        Double result = null;
+
+        if (current == null) {
+            LOG.debug("Current is null, returning");
```

Comment:
Give more context in the log message. Imagine that this is buried inside a log file, with a dozen entities that have this kind of enricher and you want to figure out from the customer's log what this refers to. I'd do something like:

```
LOG.debug("{} not updating target sensor of {}, because current is null", this, entity);
```

Same for your other log messages below.